### PR TITLE
Emit events on deletion of referenced gardener resources

### LINKED ISSUE #1473: Improve user experience during deletion of some Gardener resources
**What would you like to be added**:
In #1463 @rfranzke and me discussed the user experience during deletion of some gardener resources, like `SecretBinding` and `CloudProfile`. The case is that when these resources are referenced by other resources, e.g. `Shoots`, `Seeds`, their deletion is accepted(the deletion timestamp is set) but for the user it seems like the deletion is stuck. Basically, there is no status or event to describe that the resource `X` is referenced by `n` resources of type Y, because of which the deletion is in waiting mode.

* An Event should be the easier solution, but the event resources have a TTL, which by default is only 1 hour. So it might be possible all events to be deleted because of their retention policy and the user again will not be provided with a valid reason why the deletion is stuck.

* The other option is these resources to be extended with `Status` field, which will required API change, but it  be a time resistant solution.

**Why is this needed**:
Improve user experience.

### PR DESCRIPTION
**Pull request by timebertt**
**What this PR does / why we need it**:
gardener-controller-manager now emits Events, when the user tries to delete a resource which is still being referenced in another resource and the deletion is therefore blocked (for example a SecretBinding that is still referenced by a Shoot). This is done for the following resources:
- BackupBuckets
- CloudProfiles
- Quotas
- Seeds

**Which issue(s) this PR fixes**:
Fixes #1473

**Special notes for your reviewer**:
- Please check, if there are additional resources for which this functionality should be added
- Should I add test cases for this logic?
- I thought about adding an Event, when the user tries to delete a Secret that is still referenced by a SecretBinding or Seed. The problem is, that currently there is no controller where this logic can simply be added to. WDYT?

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

**Comment by timebertt** [comment]
cc @vpnachev

File: pkg/apis/core/v1alpha1/types_backupbucket.go [hunk 1]
```diff
@@ -82,3 +82,8 @@ type BackupBucketProvider struct {
 	// Region is the region of the bucket.
 	Region string `json:"region"`
 }
+
+const (
+	// BackupBucketEventReferenced indicates that the BackupBucket is still being referenced by at least one BackupEntry
+	BackupBucketEventReferenced = "Referenced"
```
Comment (zanetworker) [inline pkg/apis/core/v1alpha1/types_backupbucket.go]: Maybe its more convenient to define a constant called `ResourceReferenced` (could be in pkg/apis/core/v1alpha1/constants/types_constants.go)  or so and reuse it in all the required places since it will always have the same value?

File: pkg/apis/core/v1alpha1/types_cloudprofile.go [hunk 2]
```diff
@@ -158,3 +158,8 @@ type VolumeType struct {
 	// +optional
 	Usable *bool `json:"usable,omitempty"`
 }
+
+const (
+	// CloudProfileEventReferenced indicates that the CloudProfile is still being referenced by at least one Shoot
+	CloudProfileEventReferenced = "Referenced"
```
Comment (zanetworker) [inline pkg/apis/core/v1alpha1/types_cloudprofile.go]: same as above.

File: pkg/apis/core/v1alpha1/types_quota.go [hunk 3]
```diff
@@ -54,3 +54,8 @@ type QuotaSpec struct {
 	// Scope is the scope of the Quota object, either 'project' or 'secret'.
 	Scope corev1.ObjectReference `json:"scope"`
 }
+
+const (
+	// QuotaEventReferenced indicates that the Quota is still being referenced by at least one SecretBinding
+	QuotaEventReferenced = "Referenced"
```
Comment (zanetworker) [inline pkg/apis/core/v1alpha1/types_quota.go]: same as above.

File: pkg/apis/core/v1alpha1/types_secretbinding.go [hunk 4]
```diff
@@ -45,3 +45,9 @@ type SecretBindingList struct {
 	// Items is the list of SecretBindings.
 	Items []SecretBinding `json:"items"`
 }
+
+const (
+	// SecretBindingEventReferenced indicates that the SecretBinding is still being referenced
+	// by at least one other object (e.g. a Shoot)
+	SecretBindingEventReferenced = "Referenced"
```
Comment (zanetworker) [inline pkg/apis/core/v1alpha1/types_secretbinding.go]: same as above.

File: pkg/apis/core/v1alpha1/types_seed.go [hunk 5]
```diff
@@ -156,6 +156,9 @@ const (
 	// SeedTaintInvisible is a constant for a taint key on a seed that marks it as invisible. Invisible seeds
 	// are not considered by the gardener-scheduler.
 	SeedTaintInvisible = "seed.gardener.cloud/invisible"
+	// SeedEventReferenced indicates that the Seed is still being referenced by at least one other object
+	// (e.g. Shoot or BackupBucket)
+	SeedEventReferenced = "Referenced"
```
Comment (zanetworker) [inline pkg/apis/core/v1alpha1/types_seed.go]: same.

File: pkg/controllermanager/controller/cloudprofile/cloudprofile_control.go [hunk 6]
```diff
@@ -135,11 +138,11 @@ func (c *defaultControl) ReconcileCloudProfile(obj *gardencorev1alpha1.CloudProf
 			}
 			return nil
 		}
-		message := "Can't delete CloudProfile, because Shoots are still referencing it."
-		if len(associatedShoots) != 0 {
```
Comment (zanetworker) [inline pkg/controllermanager/controller/cloudprofile/cloudprofile_control.go:139]: why did you remove this check?

Comment (timebertt) [reply pkg/controllermanager/controller/cloudprofile/cloudprofile_control.go:139]: If I see that correctly, the check is redundant because if it wasn't true the method would have returned in the previous if block anyways.

Comment (zanetworker) [reply pkg/controllermanager/controller/cloudprofile/cloudprofile_control.go:139]: had to expand, to see the previous block :)

File: pkg/apis/core/v1alpha1/types_backupbucket.go (hunk 1 above)
Comment (timebertt) [reply pkg/apis/core/v1alpha1/types_backupbucket.go]: Done, PTAL :)

File: pkg/controllermanager/controller/seed/seed_control.go [hunk 7]
```diff
@@ -227,13 +228,18 @@ func (c *defaultControl) ReconcileSeed(obj *gardencorev1alpha1.Seed, key string)
...
-			seedLogger.Infof("%s Shoots=%v", parentLogMessage, associatedShoots)
+			message := fmt.Sprintf("%s Shoots=%v", parentLogMessage, associatedShoots)
+			seedLogger.Info(message)
+			c.recorder.Event(seed, corev1.EventTypeNormal, v1alpha1constants.EventResourceReferenced, message)
 		}
 		if len(associatedBackupBuckets) != 0 {
-			seedLogger.Infof("%s BackupBuckets=%v", parentLogMessage, associatedBackupBuckets)
+			message := fmt.Sprintf("%s BackupBuckets=%v", parentLogMessage, associatedBackupBuckets)
+			seedLogger.Info(message)
+			c.recorder.Event(seed, corev1.EventTypeNormal, v1alpha1constants.EventResourceReferenced, message)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_control.go:240]: Instead of sending two log messages and two events here, can we first collect and construct the message that we wanna send and then log+record it only once?

**Comment by rfranzke** [comment]
> Should I add test cases for this logic?

We don't have any tests for these things that and I guess it's hard to write them now. We plan to switch to the controller-runtime eventually, and the PR is considerably small, so IMHO it's okay as it is now.

> I thought about adding an Event, when the user tries to delete a Secret that is still referenced by a SecretBinding or Seed. The problem is, that currently there is no controller where this logic can simply be added to. WDYT?

We don't have a controller watching secrets and I don't think it makes sense to introduce one just for the sake of sending one event in this case. Thanks for considering, though.

**Comment by rfranzke** [comment]
Why is the PR marked as "WIP"? What are you planning to implement that is not yet there?

File: pkg/controllermanager/controller/seed/seed_control.go (hunk 7 above)
Comment (timebertt) [reply pkg/controllermanager/controller/seed/seed_control.go:240]: I actually thought about that, too. The reason why I reused the current structure with two logs is because of the deduplication mechanism in the Events API. As long as the related object, event type, reason and most importantly message are the same, events are aggregated into one event series. As soon as one of them (e.g. event message) changes, it will be displayed to the user as a seperate event (series). This means that, when one of the referencing objects is indeed deleted, it will disappear from the message and therefore a new event series will be displayed to the user, for example:

```
Events:
  Type    Reason              Age                  From                         Message
  ----    ------              ----                 ----                         -------
  Normal  ResourceReferenced  97s (x15 over 3m7s)  gardener-controller-manager  ControllerInstallations for seed "dev-aws" still pending, cannot release seed
  Normal  ResourceReferenced  24s (x9 over 2m23s)  gardener-controller-manager  Can't delete Seed, because the following objects are still referencing it: Shoots=[garden-dev/fake-aws-1 garden-dev/fake-aws-2] BackupBuckets=[5c7171fe-abcd-11e9-9dcd-acde48001122]
  Normal  ResourceReferenced  15s                  gardener-controller-manager  Can't delete Seed, because the following objects are still referencing it: Shoots=[garden-dev/fake-aws-1 garden-dev/fake-aws-2]
```

I think, this might be confusing to the user. Yes, the problem is actually the same, if we split it into seperate events per resource, but at least it is a little less confusing and consistent with my other changes in the deletion flow of resources that can only be referenced by one other resource type.

Actually, I am not sure, what is the right way to do it in this case.
Just wanted to give some context of my considerations. WDYT now?

**Comment by timebertt** [comment]
> We don't have a controller watching secrets and I don't think it makes sense to introduce one just for the sake of sending one event in this case.

Yeah, that was exactly my thought. Would be nice to improve the user experience on this end too, but I don't think this is enough reason to add a new controller...

> Why is the PR marked as "WIP"? What are you planning to implement that is not yet there?

Actually I wanted some feedback on whether I covered all of the resources for which events should be added or if there are additional resources for which I should also add this functionality.
But if I have already covered all resources, then we can remove the WIP label.

File: pkg/controllermanager/controller/seed/seed_control.go (hunk 7 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/seed/seed_control.go:240]: 👍 okay!

**Review by rfranzke — APPROVED** [review]
> But if I have already covered all resources, then we can remove the WIP label.

I cannot think of others atm, so I'd consider the PR finished, /lgtm.
If we see that we could have the same for also other resources we can do it then separately.

**Review by zanetworker — APPROVED** [review]
